assembly/src/assembler/module_provider.rs

```diff
@@ -0,0 +1,59 @@
+use super::{BTreeMap, Library, LibraryError, Module, ProcedureId, Vec};
+
+// MODULE PROVIDER
+// ================================================================================================
+
+/// A module provider that will store the procedure ASTs used during the compilation process.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ModuleProvider {
+    modules: Vec<Module>,
+    /// Map from procedure id to module index + proc index in the module.
+    procedures: BTreeMap<ProcedureId, usize>,
+}
+
+impl ModuleProvider {
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Fetch a module that contains the provided procedure id.
+    pub fn get_module(&self, id: &ProcedureId) -> Option<&Module> {
+        // this will panic only if there is a bug in `Self::add_module`.
+        self.procedures.get(id).map(|i| &self.modules[*i])
+    }
+
+    // MODULE AND LIBRARY MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Provide a given module.
+    ///
+    /// # Errors
+    ///
+    /// Will error if there is a duplicated module path.
+    pub fn add_module(&mut self, module: Module) -> Result<(), LibraryError> {
```

Comment:
I would change the comment to: "Adds the provided module to this module provider".

Also, I wonder if this needs to be a part of the public interface.

Reply:
I'm usually a bit more permissive with types that aren't exported as public, but it makes sense to keep the function private. Fixed!